File: src/DashSpec.Core/Parsing/ParserUtilities.cs
namespace DashSpec.Core.Parsing;

internal static class ParserUtilities
{
    public static TokenReader CreateReader(string text)
    {
        var tokens = DashSpecLexer.Tokenize(text);
        return new TokenReader(tokens);
    }

    public static IReadOnlyList<string> ParseFilterPlacementList(TokenReader reader, string blockName)
    {
        if (reader.IsAt(TokenKind.LBrace))
        {
            return PropertyBlockParser.ParseCommaListBlock(reader, blockName);
        }

        return reader.ReadCommaListInline();
    }

    /// <summary>Reads optional <c>ref &lt;id&gt;</c> postfix without crossing a newline.</summary>
    public static string? TryReadLayoutRef(TokenReader reader)
    {
        if (!reader.TryKeywordSameLine("ref"))
        {
            return null;
        }

        return reader.ReadIdentSameLine();
    }
}

File: src/DashSpec.Modeling.Parse/ParserUtilities.fs
namespace DashSpec.Modeling.Parse

open DashSpec.Modeling.Parse.Lexing

module ParserUtilities =

    let createReader (text: string) = TokenReader(DashSpecLexer.tokenize text)

    /// Reads optional ref &lt;id&gt; postfix without crossing a newline.
    let tryReadLayoutRef (reader: TokenReader) =
        if reader.TryKeywordSameLine "ref" then
            Some(reader.ReadIdentSameLine())
        else
            None

    let parseFilterPlacementList (reader: TokenReader) (endKind: string) (blockName: string) =
        if reader.IsOnNewline() then
            reader.SkipNewlines()
            PropertyBlockParser.parseCommaListBlock reader endKind blockName
        else
            reader.ReadCommaListInline()